Helm chart should allow configuring resource requests

lihaiswu opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
The Helm chart does not set nor allow setting the resource.request or resource.limit for any of the deployment objects. Each of the deployment objects should read a corresponding resources value to allow the user to set/modify the requests as appropriate to their cluster.

## Description
<!--- Describe your changes in detail -->
Move default resource values to chart/openfaas/values.yaml. User can override the value if needed.


## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
https://github.com/openfaas/faas-netes/issues/213
- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))


## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
I deployed openfaas by with -f values.yaml, and I can see the resources in deployed object use the resource defined in values.yaml.
The default gateway.resources in values.yaml is:
```
  resources:
    requests:
      memory: "120Mi"
      cpu: "50m"
```
After i copy values.yaml  to values2.yaml, and update the gateway.resources to:
```
  resources:
    requests:
      memory: "120Mi"
      cpu: "50m"
    limits:
      memory: "200Mi"
      cpu: "100m"
```
And install the helm chart with option `-f values2.yaml`, i can see the resources by `kubectl describe deployment/gateway`:
```
  Containers:
   gateway:
    Image:  openfaas/gateway:0.14.4
    Port:   8080/TCP
    Limits:
      cpu:     100m
      memory:  200Mi
    Requests:
      cpu:      50m
      memory:   120Mi
```

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

LucasRoesler commented:
@lihaiswu Overall it looks good.  

Can you provide just a little more details  about the specific values file you used for your testing? It will be helpful for historical reference.  

For validating the change, you could also run `./contrib/create-static-manifest.sh` in the project root. That script is used to generate the static yaml files from the helm chart. If the yaml files don't change, then it is a good sign.  I ran it in your branch and only saw it reorder a few keys, so it looks good to me.  But I encourage you to also give it a try. 


The only other change I am going to ask for, can you expand the table in the README that documents the available options.  There should be a new row for each new resource variable you added.  Something like this for each would perfect

```md
| `faasnetes.resource` | CPU/Memory resource requests/limits | Memory: `120Mi`, CPU: `50m` |
```

lihaiswu commented:
Thanks @LucasRoesler for the review. I've updated the readme and run `./contrib/create-static-manifest.sh`. It looks good.  Also updated the test section in the PR.

LucasRoesler commented:
Looks great.  There is only one small issue, but I think we can resolve it pretty easily.  It looks like the commit signing is working on your last commit, but unfortunately, we need the signature on all of the commits.  If you can do one more thing, then it will be ready

```sh
git reset --soft HEAD~3
git commit -s 
git push -f
```

This will squash all of your commits into a single commit, sign it, and then updates the PR.

lihaiswu commented:
@LucasRoesler thanks for your kindly help. I've squashed everything into a single commit and signing that commit based on your instructions.

alexellis commented on README.md:
```diff
@@ -46,7 +46,16 @@ faas-netes can be configured with environment variables, but for a full set of o
 | `read_timeout`      | HTTP timeout for reading the payload from the client caller (in seconds). Default: `60s`        |
-| `image_pull_policy` | Image pull policy for deployed functions (`Always`, `IfNotPresent`, `Never`.  Default: `Always` |
+| `image_pull_policy` | Image pull policy for deployed functions (`Always`, `IfNotPresent`, `Never`).  Default: `Always` |
+| `gateway.resource`        | CPU/Memory resource requests/limits (memory: `120Mi`, cpu: `50m`) |
```
Should this be `resource` or `resources`?

alexellis replied:
Pretty sure it should be `resources` looking at the Kubernetes YAML

LucasRoesler replied:
you are right, i misspelled it here https://github.com/openfaas/faas-netes/pull/462#issuecomment-507835566

lihaiswu replied:
Changed resource to resource

alexellis commented:
Can we please :pray: have an override for armhf in this PR to set Prometheus to request 125Mi instead of 512Mi? It's blocking OF on RPi since I merged the new "generate YAML" PR. RPi used to have its own set of "reasonable" resource values.

alexellis commented:
I should also mention we have a conflict on values.yaml now

@LucasRoesler is it easier for you to submit a new branch that takes @lihaiswu's commits and then adds on top so that:

A) We get this in
B) We unblock ARM
C) Hai still gets his commits + credit ?

alexellis commented:
armhf -> https://github.com/openfaas/faas-netes/blob/master/contrib/values-armhf.yaml

@lihaiswu you still have a rebase error :sob: 

Can you try fetching from upstream and rebasing (do not merge)?

Share your remote side ?
```
git remote -v
```

Alex

alexellis commented:
```
Merge remote-tracking branch 'origin/issues-213' into issues-213  …
```

What I said in my comment was: "please do not do a merge commit" :cry: 

@LucasRoesler can you help @lihaiswu out with this?

lihaiswu commented:
@alexellis is this good now? sorry :-)

lihaiswu commented:
> Can we please 🙏 have an override for armhf in this PR to set Prometheus to request 125Mi instead of 512Mi? It's blocking OF on RPi since I merged the new "generate YAML" PR. RPi used to have its own set of "reasonable" resource values.

Updated.

alexellis commented:
Thank you for working through the feedback and I hope that you are becoming more familiar with git and the pull request model. Lucas thanks for your assist

alexellis commented:
Lucas please could you bump the helm chart version and run make charts?

You can push remote branch for this so there should be no need to use your fork.

lihaiswu commented:
@alexellis @LucasRoesler thanks for helping me understand the process and merge the first PR successfully :-).
